// @apollo/server/__tests__/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const index_1 = require("../index");
const testModule = {
    typeDefs: index_1.gql `
    type Book {
      title: String
      author: String
    }

    type Query {
      books: [Book]
    }
  `,
    resolvers: {
        Query: {
            books: () => [
                {
                    title: "Harry Potter and the Chamber of Secrets",
                    author: "J.K. Rowling"
                },
                {
                    title: "Jurassic Park",
                    author: "Michael Crichton"
                }
            ]
        }
    },
};
describe("ApolloServer", () => {
    it("can execute a query", async () => {
        const operation = await (new index_1.ApolloServer({
            modules: [testModule],
        })).executeOperation({
            query: 'query GetBooks { books { author } }',
        });
        expect(operation).toHaveProperty(['data', 'books', 0, 'author']);
        expect(operation).toHaveProperty('errors', undefined);
    });
});
