// @apollo/server/execution/jsutils/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
var devAssert_1 = require("./devAssert");
Object.defineProperty(exports, "devAssert", { enumerable: true, get: function () { return devAssert_1.devAssert; } });
var inspect_1 = require("./inspect");
Object.defineProperty(exports, "inspect", { enumerable: true, get: function () { return inspect_1.inspect; } });
var invariant_1 = require("./invariant");
Object.defineProperty(exports, "invariant", { enumerable: true, get: function () { return invariant_1.invariant; } });
var isInvalid_1 = require("./isInvalid");
Object.defineProperty(exports, "isInvalid", { enumerable: true, get: function () { return isInvalid_1.isInvalid; } });
var isNullish_1 = require("./isNullish");
Object.defineProperty(exports, "isNullish", { enumerable: true, get: function () { return isNullish_1.isNullish; } });
var isObjectLike_1 = require("./isObjectLike");
Object.defineProperty(exports, "isObjectLike", { enumerable: true, get: function () { return isObjectLike_1.isObjectLike; } });
var isPromise_1 = require("./isPromise");
Object.defineProperty(exports, "isPromise", { enumerable: true, get: function () { return isPromise_1.isPromise; } });
var memoize3_1 = require("./memoize3");
Object.defineProperty(exports, "memoize3", { enumerable: true, get: function () { return memoize3_1.memoize3; } });
var nodejsCustomInspectSymbol_1 = require("./nodejsCustomInspectSymbol");
Object.defineProperty(exports, "nodejsCustomInspectSymbol", { enumerable: true, get: function () { return nodejsCustomInspectSymbol_1.nodejsCustomInspectSymbol; } });
__exportStar(require("./ObjMap"), exports);
var promiseForObject_1 = require("./promiseForObject");
Object.defineProperty(exports, "promiseForObject", { enumerable: true, get: function () { return promiseForObject_1.promiseForObject; } });
var promiseReduce_1 = require("./promiseReduce");
Object.defineProperty(exports, "promiseReduce", { enumerable: true, get: function () { return promiseReduce_1.promiseReduce; } });
